### ISSUE TITLE
Is it suitable for speech enhancement?

### ISSUE THREAD

**Issue by key2miao** [s3prl/s3prl#84]
Hello, I would like to apply upstream model into the speech enhancement downstream task. But I am not sure if it is suitable for speech enhancement. I would like to know how to reconstruct the predicted spectrogram (like filter bank) into waveform? 
Thank you very much.

**Comment by leo19941227** [comment]
Hi,

I think it depends on different pretraining methods. Some are pretrained by reconstruction loss and might be more aware of spectrogram details, while some are pretrained by negative sampling and in that case the model was never enforced to learn information necessary for a full reconstruction. But I am just guessing and you can always conduct experiments to verify this.

If we wish to reconstruct spectrogram, I would recommend reconstruct to linear spectrogram instead of filter bank, since you can just run istft to get proper waveform given the phase extracted by input waveform. This setting is quite standard for speech enhancement and typically get proper scores in pesq or stoi. While if you reconstruct to fbank, then things become complicated that you will need to depend on a good vocoder to get good waveform, and many times bad results just due to a bad vocoder. Finally, the target spectrogram, input phase for istft should all be prepared by your customized downstream dataloader.

Hope this answer your question!

**Comment by key2miao** [comment]
Thank you very much. Could you share the related materials? I see there are some funcitons in 'audio.py' and whether they can be used in this task?

**Comment by key2miao** [comment]
Hi, If I would like to train the upstream model for speech enhancement, I should use the linear feature and don't need other changes?

**Comment by leo19941227** [comment]
Hi,

1. We do not yet include materials for speech enhancement in this repo.
2. Yes. However just for reference, previously I have pretrained Mockingjay to reconstruct linear spectrogram in a self-supervised way, and the results were much worst than simply augmenting noise for input during pretraining.

**Comment by key2miao** [comment]
> Hi,
> 
> 1. We do not yet include materials for speech enhancement in this repo.
> 2. Yes. However just for reference, previously I have pretrained Mockingjay to reconstruct linear spectrogram in a self-supervised way, and the results for speech enhancement were much worst than simply augmenting noise for input during pretraining.

Thank you for your kind reply. BTW, could I know if there will be new DL/ML courses by Prof. Hung-yi Lee in this term and when the new courses will start? Thank you so much.

**Comment by sw005320** [comment]
I think this is a good idea to include speech enhancement and separation in the future. We could have some up sampling layers to recover the time resolution. In my experience, it’s working (not the best but this task does not aim to achieve the best performance for each downstream task). We actually discuss this possibility internally, but make it lower priorities for now. Stay tuned!

**Comment by andi611** [comment]
> > Hi,
> > 
> > 1. We do not yet include materials for speech enhancement in this repo.
> > 2. Yes. However just for reference, previously I have pretrained Mockingjay to reconstruct linear spectrogram in a self-supervised way, and the results for speech enhancement were much worst than simply augmenting noise for input during pretraining.
> 
> Thank you for your kind reply. BTW, could I know if there will be new DL/ML courses by Prof. Hung-yi Lee in this term and when the new courses will start? Thank you so much.

Hi, you can find all the courses given by Prof. Hung-yi Lee in this [website](https://speech.ee.ntu.edu.tw/~tlkagk/courses.html).
I will close this issue for now.
